src/passes/transform/clock_gate.cpp

```diff
@@ -0,0 +1,124 @@
+#include "coreir/passes/transform/clock_gate.h"
+
+std::string CoreIR::Passes::ClockGate::ID = "clock_gate";
+
+namespace {
+struct CEInfo {
+  bool can_replace = false;
+  CoreIR::Instance *reg = nullptr;
+  CoreIR::Instance *mux = nullptr;
+  int mux_in_port = 0;
+  CEInfo() : can_replace(false) {}
+  CEInfo(CoreIR::Instance *reg, CoreIR::Instance *mux, bool mux_in_port)
+      : can_replace(true), reg(reg), mux(mux), mux_in_port(mux_in_port) {}
+};
+
+// Checks if inst of generated module and generated module has the following
+// name
+bool instance_of(CoreIR::Instance *inst, std::string ns, std::string name) {
+  auto mod = inst->getModuleRef();
+  if (!mod->isGenerated()) {
+    return false;
+  }
+  auto gen = mod->getGenerator();
+  return gen->getName() == name && gen->getNamespace()->getName() == ns;
+}
```

Comment:
This seems like it could be generally useful enough to go into a `transforms/util.{hpp,cpp}`

Reply:
Also (nit), spacing between the function below is inconsistent with others